include/dba.php

```diff
@@ -497,6 +441,272 @@ public function any_value_fallback($sql) {
+			$sql = "/*".$a->callstack()." */ ".$sql;
+		}
+
+		switch (self::$dbo->driver) {
+			case 'pdo':
+				if (!$stmt = self::$dbo->db->prepare($sql)) {
+					$errorInfo = self::$dbo->db->errorInfo();
+					self::$dbo->error = $errorInfo[2];
+					self::$dbo->errorno = $errorInfo[1];
+					$retval = false;
+					break;
+				}
+
+				foreach ($args AS $param => $value) {
+					$stmt->bindParam($param, $args[$param]);
+				}
+
+				if (!$stmt->execute()) {
+					$errorInfo = self::$dbo->db->errorInfo();
+					self::$dbo->error = $errorInfo[2];
+					self::$dbo->errorno = $errorInfo[1];
+					$retval = false;
+				} else {
+					$retval = $stmt;
+				}
+				break;
+			case 'mysqli':
+				$stmt = self::$dbo->db->stmt_init();
+
+				if (!$stmt->prepare($sql)) {
+					self::$dbo->error = self::$dbo->db->error;
+					self::$dbo->errorno = self::$dbo->db->errno;
+					$retval = false;
+					break;
+				}
+
+				$params = '';
+				$values = array();
+				foreach ($args AS $param => $value) {
+					if (is_int($args[$param])) {
+						$params .= 'i';
+					} elseif (is_float($args[$param])) {
+						$params .= 'd';
+					} elseif (is_string($args[$param])) {
+						$params .= 's';
+					} else {
+						$params .= 'b';
+					}
+					$values[] = &$args[$param];
+				}
+
+				array_unshift($values, $params);
+
+				call_user_func_array(array($stmt, 'bind_param'), $values);
+
+				if (!$stmt->execute()) {
+					self::$dbo->error = self::$dbo->db->error;
+					self::$dbo->errorno = self::$dbo->db->errno;
+					$retval = false;
+				} else {
+					$stmt->store_result();
+					$retval = $stmt;
+				}
+				break;
+			case 'mysql':
+				// For the old "mysql" functions we cannot use prepared statements
+				foreach ($args AS $param => $value) {
+					if (is_int($args[$param]) OR is_float($args[$param])) {
+						$replace = intval($args[$param]);
+					} else {
+						$replace = "'".dbesc($args[$param])."'";
+					}
+
+					$pos = strpos($sql, '?');
+					if ($pos !== false) {
+						$sql = substr_replace($sql, $replace, $pos, 1);
+					}
+				}
+
+				$retval = mysql_query($sql, self::$dbo->db);
+				if (mysql_errno(self::$dbo->db)) {
+					self::$dbo->error = mysql_error(self::$dbo->db);
+					self::$dbo->errorno = mysql_errno(self::$dbo->db);
+				}
+				break;
+		}
+
+		$a->save_timestamp($stamp1, 'database');
+
+		if (x($a->config,'system') && x($a->config['system'], 'db_log')) {
+
+			$stamp2 = microtime(true);
+			$duration = (float)($stamp2 - $stamp1);
+
+			if (($duration > $a->config["system"]["db_loglimit"])) {
+				$duration = round($duration, 3);
+				$backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
+				@file_put_contents($a->config["system"]["db_log"], datetime_convert()."\t".$duration."\t".
+						basename($backtrace[1]["file"])."\t".
+						$backtrace[1]["line"]."\t".$backtrace[2]["function"]."\t".
+						substr($sql, 0, 2000)."\n", FILE_APPEND);
+			}
+		}
+		return $retval;
+	}
+
+	/**
+	 * @brief Executes a prepared statement
```

Comment:
Do we need to know the number of affected rows? This function is meant for things like this:

```
e("UPDATE `item` SET `unseen` = 0 WHERE `unseen` AND `uid` = ?, local_user());
```
Maybe there are affected rows, maybe not. But if there were no affected rows, because there had been no unseen items, the query itself was done successful.